File: code/packages/rust/transistors/tests/cmos_gates_test.rs
//! Tests for CMOS logic gates built from transistors.

use transistors::cmos_gates::{CMOSAnd, CMOSInverter, CMOSNand, CMOSNor, CMOSOr, CMOSXor};
use transistors::types::CircuitParams;

// =========================================================================
// CMOS Inverter Tests
// =========================================================================

#[test]
fn inverter_truth_table() {
    // NOT gate: 0->1, 1->0
    let inv = CMOSInverter::new(None, None, None);
    assert_eq!(inv.evaluate_digital(0).unwrap(), 1);
    assert_eq!(inv.evaluate_digital(1).unwrap(), 0);
}

#[test]
fn inverter_voltage_swing_high_input() {
    // Input HIGH -> output near GND
    let inv = CMOSInverter::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let result = inv.evaluate(3.3);
    assert!(result.voltage < 0.1);
}

#[test]
fn inverter_voltage_swing_low_input() {
    // Input LOW -> output near Vdd
    let inv = CMOSInverter::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let result = inv.evaluate(0.0);
    assert!(result.voltage > 3.2);
}

#[test]
fn inverter_static_power_zero() {
    // CMOS should have near-zero static power
    let inv = CMOSInverter::new(None, None, None);
    assert!(inv.static_power() < 1e-9);
}

#[test]
fn inverter_dynamic_power() {
    // Dynamic power should be positive and scale with V^2
    let inv = CMOSInverter::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let p = inv.dynamic_power(1e9, 1e-12);
    assert!(p > 0.0);
}

#[test]
fn inverter_dynamic_power_scales_with_v_squared() {
    // Halving Vdd should reduce dynamic power by ~4x
    let inv_high = CMOSInverter::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let inv_low = CMOSInverter::new(Some(CircuitParams { vdd: 1.65, ..Default::default() }), None, None);
    let p_high = inv_high.dynamic_power(1e9, 1e-12);
    let p_low = inv_low.dynamic_power(1e9, 1e-12);
    let ratio = p_high / p_low;
    assert!(ratio > 3.5 && ratio < 4.5);
}

#[test]
fn inverter_vtc_has_sharp_transition() {
    // VTC should show output snap from HIGH to LOW
    let inv = CMOSInverter::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let vtc = inv.voltage_transfer_characteristic(10);
    assert_eq!(vtc.len(), 11);
    // First point: input=0, output should be HIGH
    assert!(vtc[0].1 > 3.0);
    // Last point: input=Vdd, output should be LOW
    assert!(vtc[vtc.len() - 1].1 < 0.5);
}

#[test]
fn inverter_rejects_invalid_input() {
    // evaluate_digital should reject non-binary inputs
    let inv = CMOSInverter::new(None, None, None);
    assert!(inv.evaluate_digital(2).is_err());
}

#[test]
fn inverter_transistor_count() {
    // Inverter uses 2 transistors
    let inv = CMOSInverter::new(None, None, None);
    let result = inv.evaluate(0.0);
    assert_eq!(result.transistor_count, 2);
}

// =========================================================================
// CMOS NAND Tests
// =========================================================================

#[test]
fn nand_truth_table() {
    let nand = CMOSNand::new(None, None, None);
    assert_eq!(nand.evaluate_digital(0, 0).unwrap(), 1);
    assert_eq!(nand.evaluate_digital(0, 1).unwrap(), 1);
    assert_eq!(nand.evaluate_digital(1, 0).unwrap(), 1);
    assert_eq!(nand.evaluate_digital(1, 1).unwrap(), 0);
}

#[test]
fn nand_transistor_count() {
    let nand = CMOSNand::new(None, None, None);
    assert_eq!(nand.transistor_count(), 4);
}

#[test]
fn nand_voltage_output_high() {
    let nand = CMOSNand::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let result = nand.evaluate(0.0, 0.0);
    assert!(result.voltage > 3.0);
}

#[test]
fn nand_voltage_output_low() {
    let nand = CMOSNand::new(Some(CircuitParams { vdd: 3.3, ..Default::default() }), None, None);
    let result = nand.evaluate(3.3, 3.3);
    assert!(result.voltage < 0.5);
}

#[test]
fn nand_rejects_invalid_input() {
    let nand = CMOSNand::new(None, None, None);
    assert!(nand.evaluate_digital(2, 0).is_err());
}

// =========================================================================
// CMOS NOR Tests
// =========================================================================

#[test]
fn nor_truth_table() {
    let nor = CMOSNor::new(None, None, None);
    assert_eq!(nor.evaluate_digital(0, 0).unwrap(), 1);
    assert_eq!(nor.evaluate_digital(0, 1).unwrap(), 0);
    assert_eq!(nor.evaluate_digital(1, 0).unwrap(), 0);
    assert_eq!(nor.evaluate_digital(1, 1).unwrap(), 0);
}

#[test]
fn nor_rejects_invalid_input() {
    let nor = CMOSNor::new(None, None, None);
    assert!(nor.evaluate_digital(0, 2).is_err());
}

// =========================================================================
// CMOS AND Tests
// =========================================================================

#[test]
fn and_truth_table() {
    let and_gate = CMOSAnd::new(None);
    assert_eq!(and_gate.evaluate_digital(0, 0).unwrap(), 0);
    assert_eq!(and_gate.evaluate_digital(0, 1).unwrap(), 0);
    assert_eq!(and_gate.evaluate_digital(1, 0).unwrap(), 0);
    assert_eq!(and_gate.evaluate_digital(1, 1).unwrap(), 1);
}

#[test]
fn and_rejects_invalid_input() {
    let and_gate = CMOSAnd::new(None);
    assert!(and_gate.evaluate_digital(2, 0).is_err());
}

// =========================================================================
// CMOS OR Tests
// =========================================================================

#[test]
fn or_truth_table() {
    let or_gate = CMOSOr::new(None);
    assert_eq!(or_gate.evaluate_digital(0, 0).unwrap(), 0);
    assert_eq!(or_gate.evaluate_digital(0, 1).unwrap(), 1);
    assert_eq!(or_gate.evaluate_digital(1, 0).unwrap(), 1);
    assert_eq!(or_gate.evaluate_digital(1, 1).unwrap(), 1);
}

#[test]
fn or_rejects_invalid_input() {
    let or_gate = CMOSOr::new(None);
    assert!(or_gate.evaluate_digital(0, 2).is_err());
}

// =========================================================================
// CMOS XOR Tests
// =========================================================================

#[test]
fn xor_truth_table() {
    let xor_gate = CMOSXor::new(None);
    assert_eq!(xor_gate.evaluate_digital(0, 0).unwrap(), 0);
    assert_eq!(xor_gate.evaluate_digital(0, 1).unwrap(), 1);
    assert_eq!(xor_gate.evaluate_digital(1, 0).unwrap(), 1);
    assert_eq!(xor_gate.evaluate_digital(1, 1).unwrap(), 0);
}

#[test]
fn xor_evaluate_from_nands() {
    // NAND-based XOR should match direct XOR
    let xor_gate = CMOSXor::new(None);
    for a in 0..=1u8 {
        for b in 0..=1u8 {
            assert_eq!(
                xor_gate.evaluate_from_nands(a, b).unwrap(),
                xor_gate.evaluate_digital(a, b).unwrap()
            );
        }
    }
}

#[test]
fn xor_rejects_invalid_input() {
    let xor_gate = CMOSXor::new(None);
    assert!(xor_gate.evaluate_digital(0, 2).is_err());
}

File: code/packages/typescript/transistors/tests/cmos_gates.test.ts
/**
 * Tests for CMOS logic gates built from transistors.
 */

import { describe, it, expect } from "vitest";
import {
  CMOSInverter,
  CMOSNand,
  CMOSNor,
  CMOSAnd,
  CMOSOr,
  CMOSXor,
  CMOSXnor,
} from "../src/cmos_gates.js";

describe("CMOSInverter", () => {
  it("should implement NOT truth table", () => {
    const inv = new CMOSInverter();
    expect(inv.evaluateDigital(0)).toBe(1);
    expect(inv.evaluateDigital(1)).toBe(0);
  });

  it("input HIGH should produce output near GND", () => {
    const inv = new CMOSInverter({ vdd: 3.3 });
    const result = inv.evaluate(3.3);
    expect(result.voltage).toBeLessThan(0.1);
  });

  it("input LOW should produce output near Vdd", () => {
    const inv = new CMOSInverter({ vdd: 3.3 });
    const result = inv.evaluate(0.0);
    expect(result.voltage).toBeGreaterThan(3.2);
  });

  it("should have near-zero static power", () => {
    const inv = new CMOSInverter();
    expect(inv.staticPower).toBeLessThan(1e-9);
  });

  it("should have positive dynamic power", () => {
    const inv = new CMOSInverter({ vdd: 3.3 });
    const p = inv.dynamicPower(1e9, 1e-12);
    expect(p).toBeGreaterThan(0);
  });

  it("halving Vdd should reduce dynamic power by ~4x", () => {
    const invHigh = new CMOSInverter({ vdd: 3.3 });
    const invLow = new CMOSInverter({ vdd: 1.65 });
    const pHigh = invHigh.dynamicPower(1e9, 1e-12);
    const pLow = invLow.dynamicPower(1e9, 1e-12);
    const ratio = pHigh / pLow;
    expect(ratio).toBeGreaterThan(3.5);
    expect(ratio).toBeLessThan(4.5);
  });

  it("VTC should show sharp transition", () => {
    const inv = new CMOSInverter({ vdd: 3.3 });
    const vtc = inv.voltageTranferCharacteristic(10);
    expect(vtc.length).toBe(11);
    // First point: input=0, output should be HIGH
    expect(vtc[0][1]).toBeGreaterThan(3.0);
    // Last point: input=Vdd, output should be LOW
    expect(vtc[vtc.length - 1][1]).toBeLessThan(0.5);
  });

  it("should reject invalid digital input", () => {
    const inv = new CMOSInverter();
    expect(() => inv.evaluateDigital(2)).toThrow();
    expect(() => inv.evaluateDigital(true as unknown as number)).toThrow();
  });

  it("should report transistor count of 2", () => {
    const inv = new CMOSInverter();
    const result = inv.evaluate(0.0);
    expect(result.transistorCount).toBe(2);
  });
});

describe("CMOSNand", () => {
  it("should implement NAND truth table", () => {
    const nand = new CMOSNand();
    expect(nand.evaluateDigital(0, 0)).toBe(1);
    expect(nand.evaluateDigital(0, 1)).toBe(1);
    expect(nand.evaluateDigital(1, 0)).toBe(1);
    expect(nand.evaluateDigital(1, 1)).toBe(0);
  });

  it("should have transistor count of 4", () => {
    const nand = new CMOSNand();
    expect(nand.transistorCount).toBe(4);
  });

  it("should output HIGH voltage for (0,0)", () => {
    const nand = new CMOSNand({ vdd: 3.3 });
    const result = nand.evaluate(0.0, 0.0);
    expect(result.voltage).toBeGreaterThan(3.0);
  });

  it("should output LOW voltage for (1,1)", () => {
    const nand = new CMOSNand({ vdd: 3.3 });
    const result = nand.evaluate(3.3, 3.3);
    expect(result.voltage).toBeLessThan(0.5);
  });

  it("should reject invalid digital input", () => {
    const nand = new CMOSNand();
    expect(() => nand.evaluateDigital(2, 0)).toThrow();
  });
});

describe("CMOSNor", () => {
  it("should implement NOR truth table", () => {
    const nor = new CMOSNor();
    expect(nor.evaluateDigital(0, 0)).toBe(1);
    expect(nor.evaluateDigital(0, 1)).toBe(0);
    expect(nor.evaluateDigital(1, 0)).toBe(0);
    expect(nor.evaluateDigital(1, 1)).toBe(0);
  });

  it("should reject invalid digital input", () => {
    const nor = new CMOSNor();
    expect(() => nor.evaluateDigital(0, 2)).toThrow();
  });
});

describe("CMOSAnd", () => {
  it("should implement AND truth table", () => {
    const andGate = new CMOSAnd();
    expect(andGate.evaluateDigital(0, 0)).toBe(0);
    expect(andGate.evaluateDigital(0, 1)).toBe(0);
    expect(andGate.evaluateDigital(1, 0)).toBe(0);
    expect(andGate.evaluateDigital(1, 1)).toBe(1);
  });

  it("should reject invalid digital input", () => {
    const andGate = new CMOSAnd();
    expect(() => andGate.evaluateDigital(true as unknown as number, 0)).toThrow();
  });
});

describe("CMOSOr", () => {
  it("should implement OR truth table", () => {
    const orGate = new CMOSOr();
    expect(orGate.evaluateDigital(0, 0)).toBe(0);
    expect(orGate.evaluateDigital(0, 1)).toBe(1);
    expect(orGate.evaluateDigital(1, 0)).toBe(1);
    expect(orGate.evaluateDigital(1, 1)).toBe(1);
  });

  it("should reject invalid digital input", () => {
    const orGate = new CMOSOr();
    expect(() => orGate.evaluateDigital(-1, 0)).toThrow();
  });
});

describe("CMOSXor", () => {
  it("should implement XOR truth table", () => {
    const xorGate = new CMOSXor();
    expect(xorGate.evaluateDigital(0, 0)).toBe(0);
    expect(xorGate.evaluateDigital(0, 1)).toBe(1);
    expect(xorGate.evaluateDigital(1, 0)).toBe(1);
    expect(xorGate.evaluateDigital(1, 1)).toBe(0);
  });

  it("NAND-based XOR should match direct XOR", () => {
    const xorGate = new CMOSXor();
    for (const a of [0, 1]) {
      for (const b of [0, 1]) {
        expect(xorGate.evaluateFromNands(a, b)).toBe(
          xorGate.evaluateDigital(a, b),
        );
      }
    }
  });

  it("should reject invalid digital input", () => {
    const xorGate = new CMOSXor();
    expect(() => xorGate.evaluateDigital(0, 2)).toThrow();
  });
});

describe("CMOSXnor", () => {
  it("should implement XNOR truth table", () => {
    const xnorGate = new CMOSXnor();
    expect(xnorGate.evaluateDigital(0, 0)).toBe(1);
    expect(xnorGate.evaluateDigital(0, 1)).toBe(0);
    expect(xnorGate.evaluateDigital(1, 0)).toBe(0);
    expect(xnorGate.evaluateDigital(1, 1)).toBe(1);
  });

  it("should report transistor count of 8", () => {
    const xnorGate = new CMOSXnor();
    const result = xnorGate.evaluate(0.0, 0.0);
    expect(result.transistorCount).toBe(8);
  });

  it("should reject invalid digital input", () => {
    const xnorGate = new CMOSXnor();
    expect(() => xnorGate.evaluateDigital(2, 0)).toThrow();
    expect(() => xnorGate.evaluateDigital(0, -1)).toThrow();
  });
});
